modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EConditional.java

```diff
@@ -59,65 +65,67 @@ public AExpression getRightNode() {
     }
 
     @Override
-    Output analyze(ClassNode classNode, SemanticScope semanticScope, Input input) {
-        if (input.write) {
+    Output analyze(ClassNode classNode, SemanticScope semanticScope) {
+        if (semanticScope.getCondition(this, Write.class)) {
             throw createError(new IllegalArgumentException("invalid assignment: cannot assign a value to conditional operation [?:]"));
         }
 
-        if (input.read == false) {
+        if (semanticScope.getCondition(this, Read.class) == false) {
             throw createError(new IllegalArgumentException("not a statement: result not used from conditional operation [?:]"));
         }
 
         Output output = new Output();
 
-        Input conditionInput = new Input();
-        conditionInput.expected = boolean.class;
-        Output conditionOutput = analyze(conditionNode, classNode, semanticScope, conditionInput);
-        PainlessCast conditionCast = AnalyzerCaster.getLegalCast(classNode.getLocation(),
-                conditionOutput.actual, conditionInput.expected, conditionInput.explicit, conditionInput.internal);
 
-        Input leftInput = new Input();
-        leftInput.expected = input.expected;
-        leftInput.explicit = input.explicit;
-        leftInput.internal = input.internal;
-        Output leftOutput = analyze(leftNode, classNode, semanticScope, leftInput);
+        semanticScope.setCondition(conditionNode, Read.class);
+        semanticScope.putDecoration(conditionNode, new TargetType(boolean.class));
+        Output conditionOutput = analyze(conditionNode, classNode, semanticScope);
+        PainlessCast conditionCast = conditionNode.cast(semanticScope);
 
-        Input rightInput = new Input();
-        rightInput.expected = input.expected;
-        rightInput.explicit = input.explicit;
-        rightInput.internal = input.internal;
-        Output rightOutput = analyze(rightNode, classNode, semanticScope, rightInput);
+        semanticScope.setCondition(leftNode, Read.class);
```

Comment:
Consider renaming this to "true clause" or something more evocative.

Reply:
Done.